content/switching-to-v0-mangling-on-nightly.md

```diff
@@ -0,0 +1,190 @@
++++
+path = "2025/10/28/switching-to-v0-mangling-on-nightly"
+title = "Switching to Rust's own mangling scheme on nightly"
+authors = ["David Wood"]
+
+[extra]
+team = "the compiler team"
+team_url = "https://www.rust-lang.org/governance/teams/compiler"
++++
+
+**TL;DR:** rustc will use its own "v0" mangling scheme by default on nightly
+versions instead of the previous default scheme which re-used C++'s mangling
+scheme
+
+#### Context
+
+When Rust is compiled into object files and binaries, each item (functions,
+statics, etc) must have a globally unique  "symbol" identifying it.
+
+In C, the symbol name of a function is just the name that the function was
+defined with, such as `strcmp`. This is straightforward and easy to
+understand, but requires that each item have a globally unique name
+that doesn't overlap with any symbols from libraries that it is linked
+against. If two items had the same symbol then when the linker tried to resolve
+a symbol to an address in memory (of a function, say), then it wouldn't know
+which symbol is the correct one.
+
+Languages like Rust and C++ define "symbol mangling schemes", leveraging information
+from the type system to give each item a unique symbol name. Without this, it would be
+possible to produce clashing symbols in a variety of ways - for example, every
+instantiation of a generic or templated function (or an overload in C++), which all
+have the same name in the surface language would end up with clashing symbols; or
+the same name in different modules, such as `a::foo` and `b::foo` would have clashing
+symbols.
+
+Rust originally used a symbol mangling scheme based on the
+[Itanium ABI's name mangling scheme][itanium-mangling] used by C++ (sometimes). Over
+the years, it was extended in an inconsistent and ad-hoc way to support Rust
+features that the mangling scheme wasn't originally designed for. Rust's current legacy
+mangling scheme has a number of drawbacks:
+
+- Information about generic parameter instantiations is lost during mangling
+- It is internally inconsistent - some paths use an Itanium ABI-style encoding
+  but some don't
+- Symbol names can contain `.` characters which aren't supported on all platforms
+- Symbol names depend on compiler internals and can't be easily replicated by
+  other compilers or tools
+
```

Comment:
I don't know if its worth expanding this, but a big downside of looking like C++ symbols is that it makes it hard for tools to know whether to apply C++ or Rust demangling to the symbol when both languages are present.

Also that all the interesting stuff is encoded in an opaque hash, but that's more or less "depends on compiler internals".